src/resultset/record.c

```diff
@@ -33,15 +33,8 @@ Record* Record_FromGroup(const ResultSetHeader *resultset_header, const Group *g
     return r;
 }
 
-size_t Record_ToString(const Record *record, char **buf, size_t *buf_cap) {
-    size_t required_len = SIValue_StringConcatLen(record->values, record->len);
-    
-    if(*buf_cap < required_len) {
-        *buf = realloc(*buf, sizeof(char) * required_len);
-        *buf_cap = required_len;
-    }
-
-    return SIValue_StringConcat(record->values, record->len, *buf, *buf_cap);
+size_t Record_ToString(const Record *record, char **record_str) {
+  return SIValue_StringConcat(record->values, record->len, record_str);
```

Comment:
Record_ToString gets called for each row in the result-set therefor the intention was to reuse record_str to avoid N allocation, N being the number of rows in the final result-set.
Your new implementation allocates record_str within each call.

Please rewrite while keeping in mind memory reuse.